### ISSUE TITLE
CG diverged with large residual norm

### ISSUE THREAD

**Issue by zcyang** [cornellius-gp/gpytorch#1297]
Below is the simply the example from https://docs.gpytorch.ai/en/v1.1.1/examples/02_Scalable_Exact_GPs/SGPR_Regression_CUDA.html
with a different dataset.

```python

import math
import numpy as np
import pandas as pd
import torch
import gpytorch
from matplotlib import pyplot as plt

# Make plots inline

import urllib.request
import os
from scipy.io import loadmat
from math import floor


# this is for running the notebook in our testing framework
smoke_test = ('CI' in os.environ)

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00291/airfoil_self_noise.dat"

data = pd.read_csv(url, sep="\t")

data = data.values
X = data[:, :5]
y = data[:, 5]

X = (X - np.mean(X, axis=0))/np.std(X, axis=0)
y = (y - np.mean(y))/np.std(y)
X, y = torch.from_numpy(X), torch.from_numpy(y)
X, y = X.float(), y.float()

train_n = int(floor(0.8 * len(X)))
train_x = X[:train_n, :].contiguous()
train_y = y[:train_n].contiguous()

test_x = X[train_n:, :].contiguous()
test_y = y[train_n:].contiguous()

  if torch.cuda.is_available():
    train_x, train_y, test_x, test_y = train_x.cuda(), train_y.cuda(), test_x.cuda(), test_y.cuda()

from gpytorch.means import ConstantMean
from gpytorch.kernels import ScaleKernel, RBFKernel, InducingPointKernel
from gpytorch.distributions import MultivariateNormal

class GPRegressionModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = ConstantMean()
        self.base_covar_module = ScaleKernel(RBFKernel())
        self.covar_module = InducingPointKernel(self.base_covar_module, inducing_points=train_x[:500, :], likelihood=likelihood)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood)

if torch.cuda.is_available():
    model = model.cuda()
    likelihood = likelihood.cuda()

training_iterations = 2 if smoke_test else 1000

# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)


def train():
    for i in range(training_iterations):
        # Zero backprop gradients
        optimizer.zero_grad()
        # Get output from model
        output = model(train_x)
        # Calc loss and backprop derivatives
        loss = -mll(output, train_y)
        loss.backward()
        print('Iter %d/%d - Loss: %.3f noise: %.3f ' % (i + 1, training_iterations, loss.item(), model.likelihood.noise.item()))
        optimizer.step()
        torch.cuda.empty_cache()
        if i % 5 == 0:
            model.eval()
            likelihood.eval()
            with gpytorch.settings.max_preconditioner_size(10), torch.no_grad():
                with gpytorch.settings.max_root_decomposition_size(30), gpytorch.settings.fast_pred_var():
                    preds = model(test_x)
                    print('Test MAE: {}'.format(torch.sqrt(torch.mean((preds.mean - test_y)**2))))
        model.train()
        likelihood.train()


# See dkl_mnist.ipynb for explanation of this flag
train()

```

At the end, the CG will diverge with warnings like:

```
NumericalWarning: CG terminated in 1000 iterations with average residual norm 1.9349431991577148 which is larger than the tolerance of 0.01 specified by gpytorch.settings.cg_tolerance. If performance is affected, consider raising the maximum number of CG iterations by running code in a gpytorch.settings.max_cg_iterations(value) context.
  NumericalWarning,
```

Want to figure out the cause of the problem? thanks!

**Comment by gpleiss** [comment]
The airfoil dataset is very ill-conditioned (likely because there are very similar entries in the dataset). Consequentially, the training kernel matrix becomes singular as the the likelihood noise term gets smaller and smaller.

If you try using Cholesky instead of conjugate gradients (add the `with gpytorch.settings.max_cholesky_size(1000000):` context manager to your training loop), you'll notice that Cholesky similarly errors out due to the ill-conditioning.

The solution for ill-conditioned datasets like this is to ensure that the likelihood noise doesn't get too small. Adding a prior to the likelihood noise fixes this:

```python
likelihood = gpytorch.likelihoods.GaussianLikelihood(
    noise_prior=gpytorch.priors.SmoothedBoxPrior(0.15, 1.5, sigma=0.001)
)
```
The [SmoothedBoxPrior](https://docs.gpytorch.ai/en/v1.2.0/priors.html#smoothedboxprior) is a differentiable approximation to the uniform prior, effectively setting a lower bound on the noise.
